transcriptomics_data_service/routers/ingest.py

```diff
@@ -26,30 +26,72 @@ async def ingest(
 ):
     # Read and process rcm file
     file_bytes = rcm_file.file.read()
-    buffer = StringIO(file_bytes.decode("utf-8"))
-    rcm = {}
-    for row in csv.DictReader(buffer):
-        print(row)
-        rcm[row[GENE_ID_KEY]] = row
-    # rcm["WASH6P"]  would return something like:
-    # {'GeneID': 'WASH6P', '<BIOSAMPLE_ID_1>': '63', '<BIOSAMPLE_ID_2>: '0', ...}
-    # TODO read counts as integers
-
-    experiment_result = ExperimentResult(
-        experiment_result_id=experiment_result_id, assembly_name=assembly_name, assembly_id=assembly_id
-    )
-
+    rcm_df = _load_csv(file_bytes)
     # Perform the ingestion in a transaction
     async with db.transaction_connection() as transaction_con:
         # For each matrix: create ONE row in ExperimentResult
+        experiment_result = ExperimentResult(
+            experiment_result_id=experiment_result_id, assembly_name=assembly_name, assembly_id=assembly_id
+        )
         await db.create_experiment_result(experiment_result, transaction_con)
 
-        # TODO For EACH cell in the matrix: create one row in GeneExpression
-        # gene_expressions: list[GeneExpression]
-        # await db.create_gene_expressions(gene_expressions, transaction_con)
+        gene_expressions: list[GeneExpression] = []
+        for gene_code, row in rcm_df.iterrows():
+
+            for sample_id, raw_count in row.items():
+                gene_expression = GeneExpression(
+                    gene_code=gene_code,
+                    sample_id=sample_id,
+                    experiment_result_id=experiment_result_id,
+                    raw_count=raw_count,
+                )
+                gene_expressions.append(gene_expression)
+        await db.create_gene_expressions(gene_expressions, transaction_con)
     return
 
 
+def _load_csv(file_bytes: bytes) -> pd.DataFrame:
+    buffer = StringIO(file_bytes.decode("utf-8"))
+    buffer.seek(0)
+    try:
+        df = pd.read_csv(buffer, index_col=0, header=0)
+
+        # Check for duplicate values in index column or header
+        if df.index.duplicated().any():
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST, detail="Duplicate values found in the Gene ID column."
+            )
+        if df.columns.duplicated().any():
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST, detail="Duplicate values found in the Sample ID row."
+            )
+
+        try:
+            df = df.applymap(lambda x: int(x) if pd.notna(x) else x)
+        except ValueError as e:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail=f"Data type error: Non-integer value found: {e}",
+            )
```

Comment:
We can let the parent `try/except` handle this exception, since we already handle ValueError there

Reply:
They both handle different cases of value error. the apply map raises a value error if the something from the map is incorrect datatype. while the outervalue error handles value error from loading the df which might be some random value error in the csv like maybe mixed datatypes in a column.

Reply:
if this is the case but non-obvious, put a comment explaining it exactly as you did here.
in general if something is non-obvious in a PR and someone comments on it, it's good to put a comment. same goes for below.